### ISSUE TITLE
Validation stops improving after the 10th epoch (reproducing Scaling NMT paper)

### ISSUE THREAD

**Issue by xuanqing94** [pytorch/fairseq#1331]
Hi I have read the issues in #625 and #645 but they do not help in my case. I find checkpoint_best.pt is the same as checkpoint10.pt even though the last epoch is 50.

Here is my training script:
```
fairseq-train \
    data-bin/wmt16_en_de_bpe32k \
    --arch transformer_vaswani_wmt_en_de_big --ddp-backend=no_c10d --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 1.0e-3 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.3 --weight-decay 0.0 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 4000 \
    --fp16 \
    --update-freq 16
    --save-dir checkpoints/transformer_wmt16_ende_big
```

Here is the log in the last epoch:
| epoch 053:  62%|▌| 188/305 [08:00<04:04,  2.09s/it, loss=11.469, nll_loss=10.892, ppl=1900.69, wps=217021, ups=0, wpb=452166.037, bsz=14805.337, num_updates=15945, lr=0.000500846, gnorm=0.282, clip=0.000, oom=0.000, loss_scale=0.016, wall=37576, train_wall=33385]

I have a few questions:
1. With such a large batch size, is it still appropriate to use warmup=4000 iterations? if so it would take 4000/305 = 13 epochs to reach the peak lr, which seems too long?
2. The loss scale in the last epoch is just 0.016, it means that the loss is still quite big for fp16, does that mean we need to switch to fp32 instead?
3. With multiple Forward-Backward propagation, the gradients are accumulated (not averaged), but if the gradient scale is already multiplied due to accumulation, why increase the learning rate won't make the weights diverge?

Thanks in advance!

**Comment by myleott** [comment]
> | epoch 053: 62%|▌| 188/305 [08:00<04:04, 2.09s/it, loss=11.469, nll_loss=10.892, ppl=1900.69, wps=217021, ups=0, wpb=452166.037, bsz=14805.337, num_updates=15945, lr=0.000500846, gnorm=0.282, clip=0.000, oom=0.000, loss_scale=0.016, wall=37576, train_wall=33385]

Your loss has exploded, probably around epoch 10. You can try a different random seed or a slightly smaller learning rate.

> With such a large batch size, is it still appropriate to use warmup=4000 iterations? if so it would take 4000/305 = 13 epochs to reach the peak lr, which seems too long?

This is what we used, but a smaller warmup may work as well.

> The loss scale in the last epoch is just 0.016, it means that the loss is still quite big for fp16, does that mean we need to switch to fp32 instead?

The loss scale is so small because your loss has exploded.

> With multiple Forward-Backward propagation, the gradients are accumulated (not averaged), but if the gradient scale is already multiplied due to accumulation, why increase the learning rate won't make the weights diverge?

We average gradients across the number of workers. So if you have 8 GPUs with `--update-freq=16` then each update will use an average of 128 mini-batch gradients.

**Comment by xuanqing94** [comment]
Hi @myleott Thanks for your help! Now I changed the learning rate to 9.0e-4 and now it seems working. However, there is still a few digits away from the numbers in paper (which is 29.2, I can reach 29.02).

I trained the models for ~110 epochs, the last log entry is:
```
epoch 117:  68%|▋| 233/342 [08:37<03:27,  1.90s/it, loss=3.805, nll_loss=2.129, ppl=4.38, wps=211294, ups=0, wpb=402378.142, bsz=13181.164, num_updates=39677, lr=0.000285761, gnorm=0.205, clip=0.000, oom=0.000, loss_scale=4.000, wall=85288, train_wall=79115]
```

After doing checkpoint averaging (from last 10 epochs) and "compond word splitting", my result is 29.02. So I am wondering if this result is within the error bar or not? Is there anything else I need to do to exactly reproduce the official result?
